### PR TITLE
Clarify sections previously using "synchronize"

### PR DESCRIPTION
**Pull request by Pennycook**
ISO C++ defines a "synchronizes-with" relationship between library calls, and defines functions in terms of their "synchronization" effects. Additionally, it refers to the new std::barrier and std::latch types as "coordination mechanisms" rather than "synchronization mechanisms".

Before this PR, several usages of "synchronize" (or "synchronization") were incompatible with these definitions, or at least potentially ambiguous. For example, several sections talked about "synchronizing" host and device data, but really meant that data would be copied from one place to another. Other sections talked about work-items or different devices "synchronizing with" one another, but without reference to specific library calls.

This PR rewords several sections, as follows:

- Where "synchronize" was used to mean "wait", the specification now says that the caller "blocks" until a certain condition holds.

- Where "synchronize" was used to mean "two copies of data are made consistent", the specification now says that data is "copied".

- Where "synchronize" was used to mean "call a barrier", the specification now talks about "coordination" of work-items and/or devices.

Any remaining occurrences of "synchronize" are related to mutexes (which ISO C++ describes as a "synchronization primitive") or to atomics and fences (which ISO C++ describes as "synchronizing with" other atomics and fences).

Closes internal issue 629.

### CODE REVIEW

**Comment by Pennycook** [comment]
A quick note to reviewers: there were also a few places where "synchronize" was used in the glossary.  I was surprised that there were such lengthy descriptions for these terms in the glossary, as they could very easily get out of sync with the main text.  Rather than rewrite these glossary descriptions to match the updated text, I simplified them and pointed back to where the relevant behaviors are specified.

**Review by keryell — APPROVED** [review]
Nice *coordination* with ISO C++. ;-)

**Comment by TApplencourt** [comment]
For `c++` spec readers like me, can he have a little definition of "coordination" somewhere in the spec (or at least a link to the C++ spec)? 

Or maybe  something like:
```
Coordination ("synchronization") between the host and any devices can be expressed in the host SYCL
```

File: adoc/chapters/programming_interface.adoc
Comment (TApplencourt) [inline adoc/chapters/programming_interface.adoc:14065]: why removing `host pointers`?

File: adoc/chapters/programming_interface.adoc
Comment (TApplencourt) [inline adoc/chapters/programming_interface.adoc:19607]: Why remove all the `printf`  and the implementation detail stuff?

**Comment by Pennycook** [comment]
> For `c++` spec readers like me, can he have a little definition of "coordination" somewhere in the spec (or at least a link to the C++ spec)?
> 
> Or maybe something like:
> 
> ```
> Coordination ("synchronization") between the host and any devices can be expressed in the host SYCL
> ```

I don't think we should redefine coordination to mean synchronization.  I'm worried that will confuse people more, because they're not exactly the same.

I think this specific usage of coordination really follows the English meaning, and just means something like "getting the host and device to work together effectively".  I'm not opposed to swapping it out for a different word.  The only usage of "coordinate" that I feel strongly about keeping is the description of barriers as "a coordination mechanism".

File: adoc/chapters/programming_interface.adoc
Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc:14065]: I was trying to avoid the phrase "synchronization with host pointers".  I felt it was ambiguous whether it meant "manual synchronization [with host pointers]" or "manual [synchronization with host pointers]", and neither is good.  The former is confusing because host pointers don't provide synchronization, and the latter is potentially confusing because it's quite close to "synchronize with".

I convinced myself that the mention of "host pointers" didn't add anything to the sentence, so the simplest fix was to take it out.  If we're trying to clarify what "manual synchronization" is, I think the most important part to highlight is that we're talking about the buffer constructor which takes a mutex.  If somebody looks at the description of that constructor, they'll see how it relates to host pointers.

File: adoc/chapters/programming_interface.adoc
Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc:19607]: The sentence before this one says that:

> An instance of the SYCL `stream` class is required to output everything ... by the time that the event associated with a command group submission enters the completed state.

If we reworded the removed sentence to replace "synchronization" with "output", it would read:

> The point at which this output occurs and the member function by which this output is performed are implementation-defined.

...which just feels like repetition to me.  We've already said that output occurs at some point before the event enters the completed state.

I don't think the reference to `printf` adds anything.  (It's also confusing, because `printf` isn't a member function.  So either `printf` isn't a valid example of what the sentence before it is trying to describe, or the sentence before it is wrong.)

In hindsight, I should probably have separated these sorts of edits from the synchronization re-wording.  But it was too tempting to try and simplify things given that I was editing the sentences already.

**Comment by TApplencourt** [comment]
Oh sorry, I was not clear. I'm highly supportive of following the C++ terminology! 

It's maybe just me, but reading the new `Coordination` section, it was still not clear what `coordination` meant concretely. At least synchronization, even if it's defined loosely, I understand what that means. 
`coordination` is a  CS term I was not familiar with; where `synchronization`, I kind of understand it may means`barrier` and co.

My suggestion was just to add a sentence to try to "ease" the reading to non-CS / C++ experts. But maybe it's not the goal of the specification to do so.  I will let other people chime in.

Regardless of this, It's a nice change and improves the spec clarity which is always good!

**Review by TApplencourt — APPROVED** [review]

File: adoc/chapters/programming_interface.adoc
Comment (TApplencourt) [reply adoc/chapters/programming_interface.adoc:19607]: >In hindsight, I should probably have separated these sorts of edits from the synchronization re-wording. But it was too tempting to try and simplify things given that I was editing the sentences already.

:D I know the feeling! Your justification makes sense! I just wanted to double those tangential changes :) 

Can we replace it with something like:

>The point at which the flush is performed is implementation-defined

I agree with the removal of `printf` which was unclear to me too. Hence my original residence to removing it. Removing stuff I don't understand seam... wrong)

File: adoc/chapters/programming_interface.adoc
Comment (TApplencourt) [reply adoc/chapters/programming_interface.adoc:14065]: Agree that the term `manual synchronization` is confusing, as we don't use the term `synchronization` anymore.
If we use particularly useful, I guess we can go deep into the `particularity` and just wrote:

>This is particularly useful when users use the mutex objects on the [code]#buffer# constructors.

or something around those line

Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc:14065]: To make things even more confusing, I think we _could_ use "manual synchronization" here, because a mutex _is_ a synchronization primitive. :laughing:

But I'm not opposed to changing it to something like you suggested.  Maybe:

> This is particularly useful when used in conjunction with the [code]#buffer# constructor overloads which accept mutex objects.

What do you think?

**Comment by Pennycook** [comment]
> Oh sorry, I was not clear. I'm highly supportive of following the C++ terminology!

No need to apologize -- you were perfectly clear!  I'm not opposed to explaining what things mean, I was just trying to point out that 'coordination ("synchronization")' might leave a reader with the impression that the terms were equivalent.


> It's maybe just me, but reading the new `Coordination` section, it was still not clear what `coordination` meant concretely. At least synchronization, even if it's defined loosely, I understand what that means. `coordination` is a CS term I was not familiar with; where `synchronization`, I kind of understand it may means`barrier` and co.

You're not alone.  I hadn't seen "coordination" used to describe barriers until I saw it in C++20.

My understanding of how things are defined in C++20 is that there's a sort of hierarchy of concepts.  A barrier provides a way to manage/organize/coordinate a set of threads.  The implementation of the barrier guarantees that all calls to a function like `barrier::arrive` _happen before_ any call unblocks from `barrier::wait`, by employing _synchronization_ operations.  (You can see an example of this [here](http://eel.is/c++draft/thread.barrier#class-14)).

I deliberately went through and tried to change every usage of "synchronization" where I wasn't 100% sure it was matching the ISO C++ definition.  If we can convince ourselves that I've been too heavy-handed here, and that some functions in SYCL _do_ provide synchronization, then I'm more than happy to revisit the text.  I expected we'd have to discuss this PR for quite some time, so please keep the questions coming!


> My suggestion was just to add a sentence to try to "ease" the reading to non-CS / C++ experts. But maybe it's not the goal of the specification to do so. I will let other people chime in.

Where I'd like us to end up eventually is even closer to ISO C++, so each function has its own _Synchronization_: section that says what synchronization operations (if any) are involved.  So, the description of `group_barrier` would have a "_Synchronization_" section saying that all the work-items must arrive at the barrier before any barrier is unblocked.  Only functions with synchronization effects would have such a section.

Would that help?  I meant to write something about this in the PR description, but forgot.

File: adoc/chapters/programming_interface.adoc
Comment (TApplencourt) [reply adoc/chapters/programming_interface.adoc:14065]: Perfect! Far clearer. I finally understood the sentence! :D

**Comment by TApplencourt** [comment]
>  Would that help? I meant to write something about this in the PR description, but forgot.

Yes! Then, perfect, we can merge this as-is and eagerly wait for the next PR :).

 I guess my comment boiled down to more of a philosophical point.  I would like to keep the spec readable for mere mortals.  

Of course, clarify / precision / unambiguous terms are the critical things in a spec! No debate about it.  But if we can avoid becoming like some other specs that are totally unreadable by non-implementers, I will appreciate it. I guess I want a "SYCL Spec for Dummies" sentences from time to time :)

File: adoc/chapters/programming_interface.adoc
Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc:14065]: New wording in 19fcedc59740213a52d6f5137103463a50c4baa3.

File: adoc/chapters/programming_interface.adoc
Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc:19607]: I've restored the note about the implementation-defined behavior in aa30f4da953f806eeee247287b65f09d4fc89084.

File: adoc/chapters/programming_interface.adoc
Comment (gmlueck) [inline adoc/chapters/programming_interface.adoc:6257]:
```suggestion
blocks depends on the type of accessor.  Those
```

Can we just say this here?  I'd rather not say "copied back" because some implementations of accessors will not require memory to be copied (e.g. USM).  Note below that we say it blocks until "the data is available".

Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc:6257]: I have a weak preference for giving some indication of what condition will unblock the call.  Could we repeat "blocks until data is available" here?

Comment (gmlueck) [reply adoc/chapters/programming_interface.adoc:6257]: Yes, that sounds good.

File: adoc/chapters/architecture.adoc
Comment (Pennycook) [reply adoc/chapters/architecture.adoc]: I added something based on your second suggestion in 66488ef.

I made a very minor tweak, so that the first occurrence of "happen before" is not in quotes, because I think we should normally use phrases like "happen before" and "synchronizes with" without quotes.  I left the quotes in the second occurrence, because it helps draw attention to a specific part of the previous sentence.

Comment (gmlueck) [reply adoc/chapters/architecture.adoc]: Thanks, I like it!

File: adoc/chapters/programming_interface.adoc
Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc]: Fixed in 9bcbefb40be659fa1a8ce05c0551801188e0e0b9.

**Review by gmlueck — APPROVED** [review]

File: adoc/chapters/architecture.adoc
Comment (AerialMantis) [inline adoc/chapters/architecture.adoc:1344]: This is not related to this PR, but this should say commands rather than kernels, as this should also apply to non-kernel invocation commands such as copies.

File: adoc/chapters/programming_interface.adoc
Comment (AerialMantis) [inline adoc/chapters/programming_interface.adoc:5965]: The way this is worded suggests that data will always be copied back once the mutex is unlocked, but if the buffer would ordinarily wait on commands accessing that data on destruction it will also coordinate, and if the buffer would ordinarily not wait, then there will be no copy back.

**Review by AerialMantis — COMMENTED** [review]
Looks good, thanks! Just a few minor comments.

**Comment by Pennycook** [comment]
> Of course, clarify / precision / unambiguous terms are the critical things in a spec! No debate about it. But if we can avoid becoming like some other specs that are totally unreadable by non-implementers, I will appreciate it. I guess I want a "SYCL Spec for Dummies" sentences from time to time :)

After thinking about this some more, I've renamed the section to "Coordination _and Synchronization_", and added a paragraph that attempts to connect the two.

File: adoc/chapters/programming_interface.adoc
Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc:5965]: I don't think I understand how to fix this, primarily because I still don't think I understand how all the buffer destruction rules work.  Could you suggest alternative wording to fix this?

Comment (gmlueck) [reply adoc/chapters/programming_interface.adoc:5965]: I think we could rewrite that last sentence to read:

> The contents of `hostData` are guaranteed to reflect contents of the image when the `std::mutex` is unlocked by the runtime.

Note that this paragraph describes the `use_mutex` property for images.  I think there is a similar paragraph earlier in the spec that describes the `use_mutex` property for buffers.  That should also be changed.

Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc]: I've made both changes in d341f85357d1c102216bee74cf307dc13fdfe287.

Comment (AerialMantis) [reply adoc/chapters/programming_interface.adoc]: I think this wording makes sense, though I think we should also clarify that this only occurs if there is a host pointer to copy to.

Comment (Pennycook) [reply adoc/chapters/programming_interface.adoc]: Is there any reason to use `use_mutex` if there isn't a host pointer?  I've not used it, but my understanding from reading the spec was that it's only useful to synchronize access to `hostData`.

**Comment by gmlueck** [comment]
@Pennycook: we decided in the WG that we can merge this PR once @AerialMantis approves it (indicating he is happy with your resolution of his concerns).  I will hold off merging #481 until this happens.

It looks like there are merged conflicts that need to be resolved, so please do that also @Pennycook.

File: adoc/chapters/architecture.adoc
Comment (gmlueck) [inline adoc/chapters/architecture.adoc:564]:
```suggestion
until the action(s) represented by the event have completed.  The [code]#event#
```

I think we normal use plural grammar when using "(s)".

Comment (Pennycook) [reply adoc/chapters/architecture.adoc]: Good catch.

File: adoc/chapters/programming_interface.adoc
Comment (AerialMantis) [reply adoc/chapters/programming_interface.adoc]: Not really, though technically it is possible to create a buffer with `use_mutex` and without a host pointer and mutex would be on any host allocation for the memory, you could also use `set_final_data` to set a host pointer.

Saying that I think it's fair to assume that it's implicit that the contents of `hostData` is only reflected if a host pointer is provided, so I think the wording is fine as it is.

**Review by AerialMantis — APPROVED** [review]

**Pull request merged**
